structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py

```diff
@@ -0,0 +1,224 @@
+                y_cost_aug,
+                buffer,
+                None,
+            )
+
+        return loss_fn
+
+    def get_dtype_device(self) -> Tuple[torch.dtype, torch.device]:
+        for param in self.loss_fn.parameters():
+            dtype = param.dtype
+            device = param.device
+
+            break
+
+        return dtype, device
+
+    def get_batch_size(self, x: Any) -> int:
+        if isinstance(x, torch.Tensor):
+            return x.shape[0]
+        else:
+            raise NotImplementedError
+
+    @contextlib.contextmanager
+    def param_grad(self) -> Generator[None, None, None]:
+        trainable_modules = [self.inference_nn, self.cost_augmented_layer]
+        requires_grad_map = {}
+        for module in trainable_modules:
+            if module is not None:
+                try:
+                    for param in module.parameters():
+                        param.requires_grad = True
+                    yield
+                finally:
+                    # set the requires_grad back to false
+                    for param in module.parameters():
+                        param.requires_grad = False
+            else:  # if there is no loss_fn, we have nothing to do.
+                warnings.warn(
+                    (
+                        "There is inference_nn in gradient based inference sampler."
+                        " Are you using the right sampler?"
+                    )
+                )
+                try:
+                    yield
+                finally:
+                    pass
+
+    def forward(
+        self,
+        x: Any,
+        labels: Optional[
+            torch.Tensor
+        ] = None,  #: If given will have shape (batch, ...)
+        buffer: Dict = None,
+        **kwargs: Any,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+
+        if labels is None:
+            y_inf: torch.Tensor = self.inference_nn(x, buffer)  # (batch_size, ...)
+            return y_inf, None
+
+        if buffer is None:
+            buffer = {}
+        # switch on gradients on parameters using context manager
+        with self.param_grad():
+
+            loss_fn = self.get_loss_fn(
+                x, labels
+            )  #: Loss function will expect labels in form (batch, num_samples or 1, ...)
+
+            loss_values: List[float] = []
+            loss_values_tensors: List[torch.Tensor] = []
+            step_number = 0
+            loss_value: Union[torch.Tensor, float] = float("inf")
+
+            # we need to enable grad because if the top-level model
+            # was being called in a validation loop, the training
+            # flag will be False for all modules. This will not allow
+            # gradient based inference to progress.
+            with torch.enable_grad():
+                while not self.stopping_criteria(step_number, float(loss_value)):
+                    y_inf: torch.Tensor = self.inference_nn(x, buffer)  # (batch_size, ...)
```

Comment:
You should call `optimizer.zero_grad(set_to_none=True)` here before computing y_inf, instead of calling it in update().